### PR TITLE
Mm 30807 granular data retention scaffold

### PR DESCRIPTION
**Pull request by maxerenberg**
#### Summary
This PR creates the necessary tables, models and APIs for the granular data retention policy feature.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-30807

#### Enterprise PR link
https://github.com/mattermost/enterprise/pull/917

#### Release Note
```release-note
Added new API endpoints:
GET    /data_retention/policies
GET    /data_retention/policies_count
POST   /data_retention/policies
GET    /data_retention/policies/{policy_id}
PATCH  /data_retention/policies/{policy_id}
DELETE /data_retention/policies/{policy_id}
GET    /data_retention/policies/{policy_id}/teams
POST   /data_retention/policies/{policy_id}/teams
DELETE /data_retention/policies/{policy_id}/teams
POST   /data_retention/policies/{policy_id}/teams/search
GET    /data_retention/policies/{policy_id}/channels
POST   /data_retention/policies/{policy_id}/channels
DELETE /data_retention/policies/{policy_id}/channels
POST   /data_retention/policies/{policy_id}/channels/search
GET    /users/{user_id}/data_retention/team_policies
GET    /users/{user_id}/data_retention/channel_policies
```

### CODE REVIEW

File: store/sqlstore/retention_policy_store.go [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+	}
+	var newPolicy model.RetentionPolicyWithTeamAndChannelCounts
+	if err = txn.SelectOne(&newPolicy, rpSelectQuery, rpSelectProps); err != nil {
+		return nil, err
+	}
+	if err = txn.Commit(); err != nil {
+		return nil, err
+	}
+	return &newPolicy, nil
+}
```
Comment (sbishel) [inline store/sqlstore/retention_policy_store.go:128]: This method would be cleaner if you created the queries, just before running them. 
```
rpcInsertQuery, rpcInsertArgs = s.buildInsertRetentionPoliciesChannelsQuery(policy.Id, policy.ChannelIds)
if _, err = executePossiblyEmptyQuery(txn, rpcInsertQuery, rpcInsertArgs...); err != nil {
	return nil, err
}
```
As it is, I find it hard to follow.

File: store/sqlstore/retention_policy_store.go [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+			delete(teamIdsMap, *teamId)
+		}
+		for teamId := range teamIdsMap {
+			return store.NewErrNotFound("Team", teamId)
+		}
+	}
+	return nil
+}
+
+func (s *SqlRetentionPolicyStore) checkChannelsExist(channelIds []string) error {
```
Comment (sbishel) [inline store/sqlstore/retention_policy_store.go:156]: Same comments as above.

File: store/sqlstore/retention_policy_store.go [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+		builder = builder.Values(policyId, channelId)
+	}
+	_, err := builder.RunWith(s.GetMaster()).Exec()
+	if err != nil {
+		// check FK constraint violation
+		if pqErr, ok := err.(*pq.Error); ok && pqErr.Code == "23503" {
+			return store.NewErrNotFound("RetentionPolicy", policyId)
+		} else if mysqlErr, ok := err.(*mysql.MySQLError); ok && mysqlErr.Number == 0x5ac {
+			return store.NewErrNotFound("RetentionPolicy", policyId)
+		}
```
Comment (sbishel) [inline store/sqlstore/retention_policy_store.go:417]: Existing pattern is to use a Switch statement when determining error type. 
Also I would define error codes as constants. Easier to understand what they mean.
```
if err != nil {
	var pqErr *pq.Error
	var mysqlErr *mySql.MySQLError
	switch {
		case errors.As(err, &pqErr):
			if pqErr.Code == POSTGRES_FOREIGN_KEY_VIOLATION {
				return store.NewErrNotFound("RetentionPolicy", policyId)
			}
		...
		default:
			return err
		}

File: store/sqlstore/store.go [hunk 4]
```diff
@@ -1025,6 +1079,15 @@ func (ss *SqlStore) RemoveIndexIfExists(indexName string, tableName string) bool
 	return true
 }
 
+func IsConstraintAlreadyExistsError(err error) bool {
```
Comment (sbishel) [inline store/sqlstore/store.go:1082]: Use case statement and make error numbers constants.

**Review by sbishel — CHANGES_REQUESTED** [review]

File: store/sqlstore/retention_policy_store.go [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+		rpcDeleteArgs  []interface{}
+		rpcInsertQuery string
+		rpcInsertArgs  []interface{}
+		rptInsertQuery string
+		rptInsertArgs  []interface{}
+		rptDeleteQuery string
+		rptDeleteArgs  []interface{}
+		rpSelectQuery  string
+		rpSelectProps  map[string]interface{}
+	)
```
Comment (sbishel) [inline store/sqlstore/retention_policy_store.go:229]: Also, these variable names are hard to distinguish between each other.  How about something like - 
```
channelDeleteQuery
teamDeleteQuery
```

File: store/sqlstore/retention_policy_store.go (hunk 1 above)
Comment (maxerenberg) [reply store/sqlstore/retention_policy_store.go:128]: The reason why I built the queries before starting the transaction is so that we wouldn't have to build the queries while the transaction is running. This reduces the amount of time for which the transaction is open.
Your point about the code being hard to follow is valid though - I'll try to add more comments and meaningful variable names to make it easier.

File: store/sqlstore/retention_policy_store.go [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+	if err := s.checkTeamsExist(patch.TeamIds); err != nil {
+		return nil, err
+	}
+	if err := s.checkChannelsExist(patch.ChannelIds); err != nil {
+		return nil, err
+	}
+
+	policyUpdateQuery := ""
+	policyUpdateArgs := []interface{}{}
+	if patch.DisplayName != "" || patch.PostDuration > 0 {
```
Comment (mkraft) [inline store/sqlstore/retention_policy_store.go:223]: Do we need all of these conditionals to check if `DisplayName` and `PostDuration` are not their default values? Or can we just invoke `builder.Set` regardless?

File: store/sqlstore/retention_policy_store.go [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+	s.CreateCompositeIndexIfNotExists("IDX_RetentionPolicies_DisplayName_Id", "RetentionPolicies",
+		[]string{"DisplayName", "Id"})
+	s.CreateIndexIfNotExists("IDX_RetentionPoliciesChannels_PolicyId", "RetentionPoliciesChannels", "PolicyId")
+	s.CreateIndexIfNotExists("IDX_RetentionPoliciesTeams_PolicyId", "RetentionPoliciesTeams", "PolicyId")
+	s.CreateForeignKeyIfNotExists("RetentionPoliciesChannels", "PolicyId", "RetentionPolicies", "Id", true)
+	s.CreateForeignKeyIfNotExists("RetentionPoliciesTeams", "PolicyId", "RetentionPolicies", "Id", true)
+}
+
+// executePossiblyEmptyQuery only executes the query if it is non-empty. This helps avoid
+// having to check for MySQL, which, unlike Postgres, does not allow empty queries.
```
Comment (mkraft) [inline store/sqlstore/retention_policy_store.go:60]: Why not just not execute it the query for *both* drivers if it's empty?

File: store/sqlstore/retention_policy_store.go [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+	return txn.Exec(query, args...)
+}
+
+func (s *SqlRetentionPolicyStore) Save(policy *model.RetentionPolicyWithTeamAndChannelIds) (*model.RetentionPolicyWithTeamAndChannelCounts, error) {
+	// Strategy:
+	// 1. Insert new policy
+	// 2. Insert new channels into policy
+	// 3. Insert new teams into policy
+
+	if err := s.checkTeamsExist(policy.TeamIds); err != nil {
```
Comment (mkraft) [inline store/sqlstore/retention_policy_store.go:74]: Do we need to check the existence of the teams and channels—here in `(*SqlRetentionPolicyStore).Save` and in `(*SqlRetentionPolicyStore).Patch`—given that we have foreign keys?

File: store/sqlstore/retention_policy_store.go [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+			sq.Eq{"PolicyId": policyId},
+			inStrings("TeamId", teamIds),
+		})
+	_, err := builder.RunWith(s.GetMaster()).Exec()
+	return err
+}
+
+// RemoveOrphanedRows removes entries from RetentionPoliciesChannels and RetentionPoliciesTeams
+// where a channel or team no longer exists.
+func (s *SqlRetentionPolicyStore) RemoveOrphanedRows(limit int) (deleted int, err error) {
```
Comment (mkraft) [inline store/sqlstore/retention_policy_store.go:490]: For another PR, right?

**Review by mkraft — CHANGES_REQUESTED** [review]
Looking really good. Just some minor stuff.

File: .gitignore [hunk 10]
```diff
@@ -10,6 +10,7 @@ webapp/yarn-error.log
 mattermost.mattermost-license
 config/mattermost.mattermost-license
 config/config.json
+config/saml-idp.crt
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply .gitignore:13]: We can name it whatever we want. *.crt is better.

File: api4/data_retention.go [hunk 11]
```diff
@@ -4,21 +4,294 @@
...
+	err := c.App.RemoveChannelsFromRetentionPolicy(policyId, lst.ChannelIds)
 	if err != nil {
 		c.Err = err
 		return
 	}
 
-	w.Write([]byte(policy.ToJson()))
+	auditRec.Success()
+	w.WriteHeader(http.StatusNoContent)
 }
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply api4/data_retention.go:349]: OK.

File: store/sqlstore/retention_policy_store.go [hunk 12]
```diff
@@ -0,0 +1,529 @@
...
+	if err := s.checkTeamsExist(policy.TeamIds); err != nil {
+		return nil, err
+	}
+	if err := s.checkChannelsExist(policy.ChannelIds); err != nil {
+		return nil, err
+	}
+
+	policy.Id = model.NewId()
+
+	policyInsertQuery, policyInsertArgs, _ := s.getQueryBuilder().
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply store/sqlstore/retention_policy_store.go:83]: Given that the table/column names are constant, and the struct fields are statically typed, it shouldn't be possible for the `build()` call to fail.

File: store/sqlstore/retention_policy_store.go (hunk 6 above)
Comment (maxerenberg) [reply store/sqlstore/retention_policy_store.go:223]: We could, but it saves us from running an unnecessary UPDATE statement.

File: store/sqlstore/retention_policy_store.go (hunk 7 above)
Comment (maxerenberg) [reply store/sqlstore/retention_policy_store.go:60]: I'm not sure if I understand your question. The MySQL driver does not allow empty queries.

File: store/sqlstore/retention_policy_store.go (hunk 8 above)
Comment (maxerenberg) [reply store/sqlstore/retention_policy_store.go:74]: There are only foreign keys between the RetentionPolicies, RetentionPoliciesChannels, and RetentionPoliciesTeams tables. In particular, there are no foreign keys between the RetentionPoliciesChannels and Channels tables, nor between the RetentionPoliciesTeams and Teams tables.

File: store/sqlstore/retention_policy_store.go (hunk 9 above)
Comment (maxerenberg) [reply store/sqlstore/retention_policy_store.go:490]: Yes, this is going to be used in the wireup PR.

**Review by sbishel — COMMENTED** [review]
Very close..please address remaining comments or why they will not be updated.  Thanks.

File: api4/data_retention.go [hunk 13]
```diff
@@ -4,21 +4,294 @@
...
+	"github.com/mattermost/mattermost-server/v5/model"
 )
 
 func (api *API) InitDataRetention() {
-	api.BaseRoutes.DataRetention.Handle("/policy", api.ApiSessionRequired(getPolicy)).Methods("GET")
+	api.BaseRoutes.DataRetention.Handle("/policy", api.ApiSessionRequired(getGlobalPolicy)).Methods("GET")
+	api.BaseRoutes.DataRetention.Handle("/policies", api.ApiSessionRequired(getPolicies)).Methods("GET")
+	api.BaseRoutes.DataRetention.Handle("/policies", api.ApiSessionRequired(createPolicy)).Methods("POST")
+	api.BaseRoutes.DataRetention.Handle("/policies/{policy_id}", api.ApiSessionRequired(getPolicy)).Methods("GET")
+	api.BaseRoutes.DataRetention.Handle("/policies/{policy_id}", api.ApiSessionRequired(patchPolicy)).Methods("PATCH")
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply api4/data_retention.go]: Done.

File: store/sqlstore/retention_policy_store.go [hunk 14]
```diff
@@ -0,0 +1,521 @@
...
+	var (
+		rpInsertQuery  string
+		rpInsertArgs   []interface{}
+		rpcInsertQuery string
+		rpcInsertArgs  []interface{}
+		rptInsertQuery string
+		rptInsertArgs  []interface{}
+		rpSelectQuery  string
+		rpSelectProps  map[string]interface{}
+	)
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply store/sqlstore/retention_policy_store.go]: Done.

File: store/sqlstore/retention_policy_store.go (hunk 3 above)
Comment (maxerenberg) [reply store/sqlstore/retention_policy_store.go:425]: Done.

File: store/sqlstore/store.go (hunk 4 above)
Comment (maxerenberg) [reply store/sqlstore/store.go:1020]: Done.

File: api4/data_retention.go [hunk 15]
```diff
@@ -4,21 +4,294 @@
...
+		return
+	}
+
+	w.Write(policy.ToJson())
+}
+
+func getPolicies(c *Context, w http.ResponseWriter, r *http.Request) {
+	limit := uint64(c.Params.PerPage)
+	offset := uint64(c.Params.Page) * limit
+	countsOnly, _ := strconv.ParseBool(r.URL.Query().Get("counts_only"))
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply api4/data_retention.go]: I removed the counts_only parameter entirely since there is now a separate endpoint for retrieving the number of policies.

File: api4/data_retention.go [hunk 16]
```diff
@@ -4,21 +4,294 @@
...
+func createPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	policy, jsonErr := model.RetentionPolicyWithAppliedFromJson(r.Body)
+	if jsonErr != nil {
+		c.SetInvalidParam("policy")
+		return
+	}
+	auditRec := c.MakeAuditRecord("createPolicy", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+	auditRec.AddMeta("policy", policy)
+	if !c.IsSystemAdmin() {
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply api4/data_retention.go]: Done.

File: api4/data_retention.go [hunk 17]
```diff
@@ -4,21 +4,291 @@
...
+
+func addTeamsToPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequirePolicyId()
+	policyId := c.Params.PolicyId
+	var teamIDs []string
+	jsonErr := json.NewDecoder(r.Body).Decode(&teamIDs)
+	if jsonErr != nil {
+		c.SetInvalidParam("team_ids")
+		return
+	}
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply api4/data_retention.go:225]: `model.ArrayFromJson` masks the JSON decoding error and returns an empty (but non-nil) array, which is now valid input. If the user submits invalid JSON, I think we should be returning an error.

File: api4/data_retention.go [hunk 18]
```diff
@@ -4,21 +4,291 @@
...
+	api.BaseRoutes.DataRetention.Handle("/policies/{policy_id:[A-Za-z0-9]+}/teams", api.ApiSessionRequired(getTeamsForPolicy)).Methods("GET")
+	api.BaseRoutes.DataRetention.Handle("/policies/{policy_id:[A-Za-z0-9]+}/teams", api.ApiSessionRequired(addTeamsToPolicy)).Methods("POST")
+	api.BaseRoutes.DataRetention.Handle("/policies/{policy_id:[A-Za-z0-9]+}/teams", api.ApiSessionRequired(removeTeamsFromPolicy)).Methods("DELETE")
+	api.BaseRoutes.DataRetention.Handle("/policies/{policy_id:[A-Za-z0-9]+}/channels", api.ApiSessionRequired(getChannelsForPolicy)).Methods("GET")
+	api.BaseRoutes.DataRetention.Handle("/policies/{policy_id:[A-Za-z0-9]+}/channels", api.ApiSessionRequired(addChannelsToPolicy)).Methods("POST")
+	api.BaseRoutes.DataRetention.Handle("/policies/{policy_id:[A-Za-z0-9]+}/channels", api.ApiSessionRequired(removeChannelsFromPolicy)).Methods("DELETE")
 }
 
-func getPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+func getGlobalPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply api4/data_retention.go:32]: I'm using the MANAGE_SYSTEM permission for now. We'll have to decide at some point whether this is the correct permission to use.

File: api4/data_retention.go [hunk 19]
```diff
@@ -4,21 +4,291 @@
...
+	}
+
+	w.Write(policy.ToJson())
+}
+
+func getPolicies(c *Context, w http.ResponseWriter, r *http.Request) {
+	limit := c.Params.PerPage
+	offset := c.Params.Page * limit
+
+	policies, err := c.App.GetRetentionPolicies(offset, limit)
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply api4/data_retention.go:53]: I decided to retrieve the count at the DataRetentionInterfaceImpl layer.

File: api4/data_retention.go [hunk 20]
```diff
@@ -4,21 +4,291 @@
...
+	limit := c.Params.PerPage
+	offset := c.Params.Page * limit
+
+	teams, err := c.App.GetTeamsForRetentionPolicy(policyId, offset, limit)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	b, _ := json.Marshal(teams)
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply api4/data_retention.go:188]: Sure.

File: api4/data_retention.go [hunk 21]
```diff
@@ -4,21 +4,291 @@
...
+
+func removeTeamsFromPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequirePolicyId()
+	policyId := c.Params.PolicyId
+	var teamIDs []string
+	jsonErr := json.NewDecoder(r.Body).Decode(&teamIDs)
+	if jsonErr != nil {
+		c.SetInvalidParam("team_ids")
+		return
+	}
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply api4/data_retention.go:253]: See comment above.

File: api4/data_retention.go [hunk 22]
```diff
@@ -4,21 +4,291 @@
...
+	limit := c.Params.PerPage
+	offset := c.Params.Page * limit
+
+	channels, err := c.App.GetChannelsForRetentionPolicy(policyId, offset, limit)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	b, _ := json.Marshal(channels)
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply api4/data_retention.go:291]: Sure.

File: model/data_retention_policy.go [hunk 23]
```diff
@@ -8,20 +8,89 @@ import (
...
+	TeamId   string
+}
+
+type RetentionPolicyWithTeamAndChannelCountsList struct {
+	Policies   []*RetentionPolicyWithTeamAndChannelCounts `json:"policies"`
+	TotalCount *int                                       `json:"total_count,omitempty"`
+}
+
+func (rp *GlobalRetentionPolicy) ToJson() []byte {
+	b, _ := json.Marshal(rp)
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply model/data_retention_policy.go:52]: Actually, here I would argue against it. Here are my reasons why:
* The structs and the `ToJson` methods are defined in the same file, so we have high cohesion between the two. If someone adds some JSON-incompatible field to the struct for whatever reason, they should be able to see the ToJson method in the same file and know to update it.
* All of our database-layer structs are either stored directly in a SQL database, or are composed of such structs. This guarantees that they will never have channels, functions, complex numbers, circular references, or any other JSON-incompatible value. Even if for some reason they did, the programmer would need to use a `db` struct tag to ignore the field, in which case they would know to do the same with a `json` tag.
* Most if not all of our other database-layer `ToJson` methods ignore the JSON error completely, including for structs which are far more complex than these ones.

If we're defining a `ToJson` method right below the struct definition, I don't see a need to return an error. It just adds unnecessary `if err != nil` boilerplate code everywhere.

File: store/sqlstore/team_store.go [hunk 24]
```diff
@@ -403,6 +403,16 @@ func (s SqlTeamStore) teamSearchQuery(term string, opts *model.TeamSearch, count
 		query = query.Where(fmt.Sprintf("(Name %[1]s ? OR DisplayName %[1]s ?)", operatorKeyword), term, term)
 	}
 
+	if opts.PolicyID != nil && *opts.PolicyID != "" {
+		query = query.
+			InnerJoin("RetentionPoliciesTeams ON t.Id = RetentionPoliciesTeams.TeamId").
+			Where("RetentionPoliciesTeams.PolicyId = ?", *opts.PolicyID)
```
Comment (mkraft) [inline store/sqlstore/team_store.go:409]:
```suggestion
			Where(sq.Eq{"RetentionPoliciesTeams.PolicyId": *opts.PolicyID})
```

File: api4/channel.go [hunk 25]
```diff
@@ -1020,18 +1025,20 @@ func searchAllChannels(c *Context, w http.ResponseWriter, r *http.Request) {
 	includeDeleted, _ := strconv.ParseBool(r.URL.Query().Get("include_deleted"))
 	includeDeleted = includeDeleted || props.IncludeDeleted
 
+	// PolicyID is ignored since that may only be used through the /data_retention/policies endpoint
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply api4/channel.go:1028]: We don't want to allow clients to use the `PolicyID` field via this API endpoint. The `model.ChannelSearch` struct is shared between both `/channels/search` and `/data_retention/policies/<policy_id>/channels/search`.

On a side note, we actually have this problem in a few other places, where API endpoints are accepting parameters which aren't officially documented because of the way we're sharing structs.

File: store/sqlstore/retention_policy_store.go [hunk 26]
```diff
@@ -0,0 +1,529 @@
...
+		tableT.SetKeys(false, "TeamId")
+		tableT.ColMap("PolicyId").SetMaxSize(26)
+		tableT.ColMap("TeamId").SetMaxSize(26)
+	}
+
+	return s
+}
+
+func (s *SqlRetentionPolicyStore) createIndexesIfNotExists() {
+	s.CreateCompositeIndexIfNotExists("IDX_RetentionPolicies_DisplayName_Id", "RetentionPolicies",
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply store/sqlstore/retention_policy_store.go:51]: When we retrieve the list of all policies, we want them to be sorted (so that the client can paginate using `page` and `per_page`). The policies will be sorted on display name and ID.

File: store/sqlstore/retention_policy_store.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (maxerenberg) [reply store/sqlstore/retention_policy_store.go:83]: On second thought, it would be better to check for the error.

File: app/app_iface.go [hunk 27]
```diff
@@ -356,6 +356,7 @@ type AppIface interface {
 	AccountMigration() einterfaces.AccountMigrationInterface
 	ActivateMfa(userID, token string) *model.AppError
 	AddChannelMember(userID string, channel *model.Channel, userRequestorId string, postRootId string) (*model.ChannelMember, *model.AppError)
+	AddChannelsToRetentionPolicy(policyId string, channelIds []string) *model.AppError
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply app/app_iface.go:358]: Sure.

File: model/data_retention_policy.go [hunk 28]
```diff
@@ -8,20 +8,125 @@ import (
...
+	TeamCount    int64 `json:"team_count"`
+}
+
+type RetentionPolicyChannel struct {
+	PolicyId  string
+	ChannelId string
+}
+
+type RetentionPolicyTeam struct {
+	PolicyId string
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply model/data_retention_policy.go:42]: Sure.

File: store/sqlstore/channel_store.go [hunk 29]
```diff
@@ -2674,6 +2674,16 @@ func (s SqlChannelStore) channelSearchQuery(term string, opts store.ChannelSearc
 		query = query.Offset(uint64(*opts.Page * *opts.PerPage))
 	}
 
+	if opts.PolicyID != "" {
+		query = query.
+			InnerJoin("RetentionPoliciesChannels ON c.Id = RetentionPoliciesChannels.ChannelId").
+			Where("RetentionPoliciesChannels.ChannelId = ?", opts.PolicyID)
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply store/sqlstore/channel_store.go:2680]: Is there any advantage to using the sq.Eq map? For one-parameter clauses, I think it's actually more readable to use `= ?`.

**Review by sbishel — APPROVED** [review]

**Review by sbishel — CHANGES_REQUESTED** [review]

File: api4/user.go [hunk 30]
```diff
@@ -92,6 +92,9 @@ func (api *API) InitUser() {
 
 	api.BaseRoutes.User.Handle("/uploads", api.ApiSessionRequired(getUploadsForUser)).Methods("GET")
 
+	api.BaseRoutes.User.Handle("/data_retention/team_policies", api.ApiSessionRequired(getTeamPoliciesForUser)).Methods("GET")
+	api.BaseRoutes.User.Handle("/data_retention/channel_policies", api.ApiSessionRequired(getChannelPoliciesForUser)).Methods("GET")
+
```
[the remark this answers is not in the document]
Comment (maxerenberg) [reply api4/user.go]: I moved them to data_retention.go.

**Review by sbishel — APPROVED** [review]

**Review by sbishel — APPROVED** [review]

**Comment by furqanmlk** [comment]
**1. Policy is taking space as a name**
image



**2. Discard Changes dialog shows up, even after Saving changes**
image



**3. Text merging on resizing the browser**
image


**4. Button text and Table Column name has mismatch cases**
image


**5. There is no edit options (three dots) on resized window**
image


**6. Error text does not match with Figma**
image


**7. 
         i. Add days in policy > Click Save > Shows "Policy name can't be blank." error
        ii.  Give the policy name > Click Save > still shows "Policy name can't be blank." error
image


**8. 0 days error is not highlighted**
image

**9. Spelling mistake**
image


**10. Data Retention Policy and Channel messages does not match in figma**
image

**Review by mkraft — APPROVED** [review]

**Comment by maxerenberg** [comment]
> Looking at the designs I'm seeing pagination on that UI which means the frontend will need to always make 2 requests.

Currently I'm just performing two SQL queries for each HTTP request. I think the overhead for this is fairly small since realistically, an organization is unlikely to create a very large number of policies.

**Comment by maxerenberg** [comment]
> **1. Policy is taking space as a name**
> image

OK so apparently there a few other entities which can accept whitespace as their display name, including teams:
Screen Shot 2021-04-06 at 2 33 49 PM

For teams, this is disabled on the UI side, but it's still allowed via the API:
```sh
$ curl -i -H "Authorization: Bearer $(cat system-token.txt)" -X PUT -d '{"display_name":" "}' http://localhost:8065/api/v4/teams/jeizk7j96fdpxpwtppe5tq1ghw/patch
HTTP/1.1 200 OK
Content-Type: application/json
Vary: Accept-Encoding
X-Request-Id: cec1ewesktbhpk8drueftdz3yc
X-Version-Id: 5.34.0.dev.1185a8fb9e6487b012c11242d989597d.true
Date: Tue, 06 Apr 2021 18:33:05 GMT
Content-Length: 386

{"id":"jeizk7j96fdpxpwtppe5tq1ghw","create_at":1615567560106,"update_at":1617733985350,"delete_at":0,"display_name":" ","name":"ad-1","description":"et iste illum reprehenderit aliquid in rem itaque in maxime eius.","email":"","type":"O","company_name":"","allowed_domains":"","invite_id":"ssd98uizmbn9fc3abjksj9bsfo","allow_open_invite":false,"scheme_id":null,"group_constrained":null}
```
@mkraft  @sbishel Do we want to prevent people from creating policies whose display names are whitespace? If so, should this only be a client-side restriction, or should it also be enforced on the backend?

**Comment by sbishel** [comment]
> @mkraft @sbishel Do we want to prevent people from creating policies whose display names are whitespace? If so, should this only be a client-side restriction, or should it also be enforced on the backend?

Yes, we should prevent this. I can't think of a reason to support blank names. It should be enforced on the backend.

**Comment by sbishel** [comment]
@furqanmlk - Some of the bugs you show appear to be UI bugs. Please make sure they get reported to @BenCookie95, I'm not sure he is paying attention to this PR.

**Comment by BenCookie95** [comment]
> @furqanmlk - Some of the bugs you show appear to be UI bugs. Please make sure they get reported to @BenCookie95, I'm not sure he is paying attention to this PR.

I see the bugs @furqanmlk has mentioned. I'll take care of them

**Review by furqanmlk — APPROVED** [review]
All API points tested and added to test plan

**Comment by agnivade** [comment]
This was merged without https://github.com/mattermost/mattermost-api-reference/pull/603. I am going to revert this.
